premerge/post-submit-testing.md

```diff
@@ -0,0 +1,169 @@
+# Post Submit Testing
+
+## Introduction
+
+While this infrastructure is focused on premerge testing, it is also important
+to make sure that the specific configuration we are testing is tested post
+commit as well. This document outlines the motivation for the need to test this
+configuration post commit, why we are utilizing this design over others, and
+how we plan on implementing this to ensure we get fast feedback scalably.
+
+## Background/Motivation
+
+LLVM has two types of testing upstream: premerge and postcommit. The premerge
+testing is performed using Github Actions every time a PR is updated before it
+is merged. Premerge testing is performed using this infrastructure (specifically
+the `./premerge` folder in llvm-zorg). Landing a PR consists of squashing the
+changes into a single commit and adding that commit to the `main` branch in the
+monorepo. We care specifically about the state of the `main` branch because it
+is what the community considers the canonical tree. Currently, commits can also
+be added to the `main` branch by directly pushing to the main branch. After a
+new commit lands in the `main` branch, postcommit testing is performed. Most
+postcommit testing is performed through the Buildbot infrastructure. The main
+Buildbot instance for LLVM has a web instance hosted at
+[lab.llvm.org](https:/lab.llvm.org). When a new commit lands in `main` the
+Buildbot instance (sometimes referred to as the Buildbot master) will trigger
+many different build configurations. These configurations are defined in the
+llvm-zorg repository under the `buildbot/` folder. These configurations are run
+on Buildbot workers that are hosted by the community.
+
+It is important that we test the premerge configuration postcommit as well. We
+need to be able to determine the state of `main` (in terms of whether the build
+passed/failed and what tests failed, if any) for certain types of automation.
+Postcommit testing enables easily checking the state of `main` at any given point
+in time. This data is crucial for figuring out which commit to revert/fix
+forward to get everything back to green. Having information on the state of
+`main` is also important for certain kinds of automation, like the planned
+premerge testing advisor that will let contributors know if tests failing in
+their PR are already failing at `main` and that it should be safe to merge
+despite the given failures.
```

Comment:
These are some very large paragraphs. The content is logical but a line break between each "point" made would make it easier to read through.